cli/src/commands/squash.rs

```diff
@@ -128,14 +130,48 @@ pub(crate) fn cmd_squash(
         &destination,
         matcher.as_ref(),
         &diff_selector,
-        description,
+        SquashedDescription::from_args(args)?,
         args.revision.is_none() && args.from.is_empty() && args.into.is_none(),
         &args.paths,
     )?;
     tx.finish(ui, tx_description)?;
     Ok(())
 }
 
+// TODO(#2882): Remove public visibility once `jj move` is deleted.
+pub(crate) enum SquashedDescription {
+    // Use this exact description.
+    Exact(String),
+    // Use the destination's description and discard the descriptions of the
+    // source revisions.
+    UseDestination,
+    // Combine the descriptions of the source and destination revisions.
+    Combine,
+}
+
+// TODO(#2882): Remove public visibility once `jj move` is deleted.
+impl SquashedDescription {
+    pub(crate) fn from_args(args: &SquashArgs) -> Result<Self, CommandError> {
+        // These options are incompatible. Clap is configured to prevent this, but we
+        // should still check.
+        if args.use_destination_message && !args.message_paragraphs.is_empty() {
+            return Err(cli_error(
+                "the argument '--message <MESSAGE>' cannot be used with \
+                 '--use-destination-message'",
+            ));
+        }
```

Comment:
nit: an `assert!` is fine IMO (and just return `Self`)

Reply:
Sounds good.